devito/interfaces.py

```diff
@@ -1,81 +1,257 @@
+            self._cache_put(self)
+
+    @classmethod
+    def indices(cls, shape):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        """
+        _indices = [t, x, y, z]
+        return _indices[:len(shape) + 1]
 
     def _allocate_memory(self):
-        super(TimeData, self)._allocate_memory()
-        if self.pad_time is True:
-            self.pointer = self.pointer[self.time_order:, :, :]
+        super(self.__class__, self)._allocate_memory()
+        if self.pad_time:
+            self._data = self._data[self.time_order:, :, :]
 
-    def __new__(cls, name, spc_shape, time_dim, time_order, save, dtype, pad_time=False):
-        if save:
-            time_dim = time_dim + time_order
-        else:
-            time_dim = time_order + 1
-        shape = tuple((time_dim,) + spc_shape)
-        return IndexedBase.__new__(cls, name, shape=shape)
+    @property
+    def forward(self):
+        """Symbol for the time-forward state of the function"""
+        return self.subs(t, t + s)
+
+    @property
+    def dt(self):
+        """Symbol for the first derivative wrt the time dimension"""
+        return as_finite_diff(self.diff(t), [t, t + s])
 
 
 class PointData(DenseData):
-    """This class is expected to eventually evolve into a full-fledged
-    sparse data container. For now, the naming follows the use in the
-    current problem.
+    """Data object for sparse point data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param point: Number of points to sample
+    :param nt: Size of the time dimension for point data
+    :param dtype: Data type of the buffered data
+
+    Note: This class is expected to eventually evolve into a
+    full-fledged sparse data container. For now, the naming and
+    symbolic behaviour follows the use in the current problem.
     """
-    def __init__(self, name, npoints, nt, dtype):
-        self.npoints = npoints
-        self.nt = nt
-        super(PointData, self).__init__(name, (nt, npoints), dtype)
 
-    def __new__(cls, name, npoints, nt, *args):
-        return IndexedBase.__new__(cls, name, shape=(nt, npoints))
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            # Initialise instance from symbol cache
+            SymbolicData.__init__(self)
+            return
+        else:
+            self.nt = kwargs.get('nt')
+            self.npoint = kwargs.get('npoint')
+            kwargs['shape'] = (self.nt, self.npoint)
+            DenseData.__init__(self, *args, **kwargs)
+            # Store final instance in symbol cache
+            self._cache_put(self)
+
+    def __new__(cls, *args, **kwargs):
+        nt = kwargs.get('nt')
+        npoint = kwargs.get('npoint')
+        kwargs['shape'] = (nt, npoint)
+        return DenseData.__new__(cls, *args, **kwargs)
+
+    @classmethod
+    def indices(cls, shape):
```

Comment:
This doesn't seem right to me. Can't think of what the correct indices would be for a sparse object but this doesn't seem useful in `SourceLike` below and that's where this object is being used.